File: src/app/static/src/crypto.ts
// Client-side crypto primitives. Must match server-side constants in src/app/crypto.py.
//
// Parameters:
//   PBKDF2-SHA256, 600 000 iterations, 256-bit master key
//   HKDF-SHA256 sub-keys with empty salt and string info
//   AES-GCM 256-bit with 12-byte random nonces
//
// Architecture:
//   MK = PBKDF2(password, salt)
//   authKey  = HKDF(MK, "trackhub-auth")           → sent to server (hex)
//   KEK      = HKDF(MK, "trackhub-enc")            → wraps DEK client-only
//   recoveryKEK      = HKDF(recoveryKey, "trackhub-recovery")
//   recoveryAuthKey  = HKDF(recoveryKey, "trackhub-recovery-auth")  → sent during recovery
//   DEK is random 256-bit, generated once at first password set.
//   Records are encrypted as nonce(12) || AES-GCM(DEK, JSON(record)).

const KDF_ITERATIONS = 600_000;

// ===== Hex =====

export function toHex(bytes: Uint8Array): string {
  let out = "";
  for (const b of bytes) out += b.toString(16).padStart(2, "0");
  return out;
}

export function fromHex(hex: string): Uint8Array {
  if (hex.length % 2) throw new Error("invalid hex length");
  const out = new Uint8Array(hex.length / 2);
  for (let i = 0; i < out.length; i++) {
    out[i] = parseInt(hex.substr(i * 2, 2), 16);
  }
  return out;
}

export function randomBytes(n: number): Uint8Array {
  const a = new Uint8Array(n);
  crypto.getRandomValues(a);
  return a;
}

// ===== KDF =====

async function pbkdf2(password: string, salt: Uint8Array, iterations: number, keyLen: number): Promise<Uint8Array> {
  const passKey = await crypto.subtle.importKey(
    "raw",
    new TextEncoder().encode(password) as BufferSource,
    { name: "PBKDF2" },
    false,
    ["deriveBits"],
  );
  const bits = await crypto.subtle.deriveBits(
    { name: "PBKDF2", salt: salt as BufferSource, iterations, hash: "SHA-256" },
    passKey,
    keyLen * 8,
  );
  return new Uint8Array(bits);
}

async function hkdf(ikm: Uint8Array, info: string, length: number): Promise<Uint8Array> {
  const ikmKey = await crypto.subtle.importKey(
    "raw",
    ikm as BufferSource,
    { name: "HKDF" },
    false,
    ["deriveBits"],
  );
  const bits = await crypto.subtle.deriveBits(
    {
      name: "HKDF",
      hash: "SHA-256",
      salt: new Uint8Array(32),
      info: new TextEncoder().encode(info),
    },
    ikmKey,
    length * 8,
  );
  return new Uint8Array(bits);
}

// ===== Public derivation API =====

export interface DerivedKeys {
  authKeyHex: string;   // sent to server
  kek: Uint8Array;      // wraps DEK
}

export async function deriveKeysFromPassword(password: string, salt: Uint8Array): Promise<DerivedKeys> {
  const mk = await pbkdf2(password, salt, KDF_ITERATIONS, 32);
  const authKey = await hkdf(mk, "trackhub-auth", 32);
  const kek = await hkdf(mk, "trackhub-enc", 32);
  return { authKeyHex: toHex(authKey), kek };
}

export async function deriveRecoveryKeys(recoveryKey: Uint8Array): Promise<{ kek: Uint8Array; authKeyHex: string }> {
  const kek = await hkdf(recoveryKey, "trackhub-recovery", 32);
  const authKey = await hkdf(recoveryKey, "trackhub-recovery-auth", 32);
  return { kek, authKeyHex: toHex(authKey) };
}

// ===== AES-GCM wrap/unwrap =====

async function importAesKey(raw: Uint8Array): Promise<CryptoKey> {
  return crypto.subtle.importKey("raw", raw as BufferSource, { name: "AES-GCM" }, false, ["encrypt", "decrypt"]);
}

export async function wrapBytes(plaintext: Uint8Array, key: Uint8Array): Promise<Uint8Array> {
  const aesKey = await importAesKey(key);
  const nonce = randomBytes(12);
  const ct = new Uint8Array(
    await crypto.subtle.encrypt({ name: "AES-GCM", iv: nonce as BufferSource }, aesKey, plaintext as BufferSource),
  );
  const out = new Uint8Array(nonce.length + ct.length);
  out.set(nonce, 0);
  out.set(ct, nonce.length);
  return out;
}

export async function unwrapBytes(blob: Uint8Array, key: Uint8Array): Promise<Uint8Array> {
  if (blob.length < 12 + 16) throw new Error("blob too short");
  const nonce = blob.subarray(0, 12);
  const ct = blob.subarray(12);
  const aesKey = await importAesKey(key);
  return new Uint8Array(
    await crypto.subtle.decrypt({ name: "AES-GCM", iv: nonce as BufferSource }, aesKey, ct as BufferSource),
  );
}

// ===== Record encryption =====

export interface EncryptedBlob {
  nonceHex: string;
  ciphertextHex: string;
}

export async function encryptRecord(obj: unknown, dek: Uint8Array): Promise<EncryptedBlob> {
  const pt = new TextEncoder().encode(JSON.stringify(obj));
  const aesKey = await importAesKey(dek);
  const nonce = randomBytes(12);
  const ct = new Uint8Array(
    await crypto.subtle.encrypt({ name: "AES-GCM", iv: nonce as BufferSource }, aesKey, pt as BufferSource),
  );
  return { nonceHex: toHex(nonce), ciphertextHex: toHex(ct) };
}

export async function decryptRecord<T = unknown>(blob: EncryptedBlob, dek: Uint8Array): Promise<T> {
  const aesKey = await importAesKey(dek);
  const nonce = fromHex(blob.nonceHex);
  const ct = fromHex(blob.ciphertextHex);
  const pt = new Uint8Array(
    await crypto.subtle.decrypt({ name: "AES-GCM", iv: nonce as BufferSource }, aesKey, ct as BufferSource),
  );
  return JSON.parse(new TextDecoder().decode(pt)) as T;
}

// ===== Bootstrap helpers (first password set) =====

export function genDEK(): Uint8Array {
  return randomBytes(32);
}

export function genRecoveryKey(): Uint8Array {
  return randomBytes(32);
}

/** Format recovery key as 8 groups of 8 hex chars, space-separated, for easy transcription. */
export function formatRecoveryKey(key: Uint8Array): string {
  const hex = toHex(key);
  const groups: string[] = [];
  for (let i = 0; i < hex.length; i += 8) groups.push(hex.slice(i, i + 8));
  return groups.join(" ");
}

export function parseRecoveryKey(input: string): Uint8Array {
  const clean = input.replace(/\s+/g, "").toLowerCase();
  if (!/^[0-9a-f]{64}$/.test(clean)) {
    throw new Error("Recovery key must be 64 hex characters");
  }
  return fromHex(clean);
}

File: android/app/src/main/kotlin/com/trackhub/crypto/Crypto.kt
package com.trackhub.crypto

import java.security.SecureRandom
import javax.crypto.Cipher
import javax.crypto.Mac
import javax.crypto.SecretKeyFactory
import javax.crypto.spec.GCMParameterSpec
import javax.crypto.spec.PBEKeySpec
import javax.crypto.spec.SecretKeySpec

/**
 * Mirror of src/app/static/src/crypto.ts (web client). Must produce byte-for-byte
 * identical output so encrypted records interop.
 *
 * Algorithms:
 *   MK = PBKDF2-SHA256(password, salt, 600_000) → 32 bytes
 *   subkey = HKDF-SHA256(MK, salt=zeros[32], info, 32 bytes)
 *   AES-GCM with 12-byte random nonce, 128-bit tag, no AAD
 *
 * Wrapped blob layout: nonce(12) || ciphertext || tag(16)
 */
object Crypto {

    private const val KDF_ITERATIONS = 600_000
    private const val GCM_TAG_BITS = 128
    private const val NONCE_LEN = 12
    private const val SALT_LEN = 16
    private const val DEK_LEN = 32

    private val random = SecureRandom()

    // ===== Hex =====

    fun toHex(bytes: ByteArray): String =
        buildString(bytes.size * 2) {
            for (b in bytes) {
                val v = b.toInt() and 0xff
                append(HEX[v ushr 4])
                append(HEX[v and 0x0f])
            }
        }

    fun fromHex(hex: String): ByteArray {
        require(hex.length % 2 == 0) { "invalid hex length" }
        val out = ByteArray(hex.length / 2)
        for (i in out.indices) {
            out[i] = ((digit(hex[i * 2]) shl 4) or digit(hex[i * 2 + 1])).toByte()
        }
        return out
    }

    private fun digit(c: Char): Int = when (c) {
        in '0'..'9' -> c - '0'
        in 'a'..'f' -> c - 'a' + 10
        in 'A'..'F' -> c - 'A' + 10
        else -> throw IllegalArgumentException("bad hex char: $c")
    }

    private val HEX = "0123456789abcdef".toCharArray()

    fun randomBytes(n: Int): ByteArray = ByteArray(n).also { random.nextBytes(it) }

    fun generateSalt(): ByteArray = randomBytes(SALT_LEN)
    fun generateDek(): ByteArray = randomBytes(DEK_LEN)
    fun generateRecoveryKey(): ByteArray = randomBytes(32)

    // ===== KDF =====

    fun pbkdf2(password: String, salt: ByteArray, iterations: Int = KDF_ITERATIONS, length: Int = 32): ByteArray {
        val spec = PBEKeySpec(password.toCharArray(), salt, iterations, length * 8)
        val factory = SecretKeyFactory.getInstance("PBKDF2WithHmacSHA256")
        return factory.generateSecret(spec).encoded
    }

    /** HKDF-SHA256 with empty 32-byte salt, matching client. */
    fun hkdf(ikm: ByteArray, info: String, length: Int = 32): ByteArray {
        val salt = ByteArray(32) // zero-filled, same as client
        val mac = Mac.getInstance("HmacSHA256")
        // Extract
        mac.init(SecretKeySpec(salt, "HmacSHA256"))
        val prk = mac.doFinal(ikm)
        // Expand
        mac.init(SecretKeySpec(prk, "HmacSHA256"))
        val infoBytes = info.toByteArray(Charsets.UTF_8)
        val out = ByteArray(length)
        var t = ByteArray(0)
        var pos = 0
        var i = 1
        while (pos < length) {
            mac.reset()
            mac.update(t)
            mac.update(infoBytes)
            mac.update(byteArrayOf(i.toByte()))
            t = mac.doFinal()
            val take = minOf(t.size, length - pos)
            System.arraycopy(t, 0, out, pos, take)
            pos += take
            i++
        }
        return out
    }

    data class DerivedKeys(val authKeyHex: String, val kek: ByteArray)

    fun deriveKeysFromPassword(password: String, salt: ByteArray): DerivedKeys {
        val mk = pbkdf2(password, salt)
        val authKey = hkdf(mk, "trackhub-auth")
        val kek = hkdf(mk, "trackhub-enc")
        return DerivedKeys(toHex(authKey), kek)
    }

    data class RecoveryKeys(val kek: ByteArray, val authKeyHex: String)

    fun deriveRecoveryKeys(recoveryKey: ByteArray): RecoveryKeys {
        val kek = hkdf(recoveryKey, "trackhub-recovery")
        val auth = hkdf(recoveryKey, "trackhub-recovery-auth")
        return RecoveryKeys(kek, toHex(auth))
    }

    // ===== AES-GCM =====

    /** Encrypt `plaintext` with `key` and given `nonce` (12 bytes). Returns ciphertext||tag. */
    fun aesGcmEncrypt(plaintext: ByteArray, key: ByteArray, nonce: ByteArray): ByteArray {
        require(nonce.size == NONCE_LEN) { "nonce must be 12 bytes" }
        val cipher = Cipher.getInstance("AES/GCM/NoPadding")
        cipher.init(Cipher.ENCRYPT_MODE, SecretKeySpec(key, "AES"), GCMParameterSpec(GCM_TAG_BITS, nonce))
        return cipher.doFinal(plaintext)
    }

    /** Decrypt `ciphertext||tag` with `key` and `nonce`. Returns plaintext. */
    fun aesGcmDecrypt(ciphertextWithTag: ByteArray, key: ByteArray, nonce: ByteArray): ByteArray {
        require(nonce.size == NONCE_LEN) { "nonce must be 12 bytes" }
        val cipher = Cipher.getInstance("AES/GCM/NoPadding")
        cipher.init(Cipher.DECRYPT_MODE, SecretKeySpec(key, "AES"), GCMParameterSpec(GCM_TAG_BITS, nonce))
        return cipher.doFinal(ciphertextWithTag)
    }

    /** Wrap a value: returns `nonce(12) || ciphertext||tag`. */
    fun wrapBytes(plaintext: ByteArray, key: ByteArray): ByteArray {
        val nonce = randomBytes(NONCE_LEN)
        val ct = aesGcmEncrypt(plaintext, key, nonce)
        return nonce + ct
    }

    /** Unwrap `nonce(12) || ciphertext||tag` back to plaintext. */
    fun unwrapBytes(blob: ByteArray, key: ByteArray): ByteArray {
        require(blob.size > NONCE_LEN + 16) { "wrapped blob too short" }
        val nonce = blob.sliceArray(0 until NONCE_LEN)
        val ct = blob.sliceArray(NONCE_LEN until blob.size)
        return aesGcmDecrypt(ct, key, nonce)
    }

    // ===== Record-shaped encryption =====

    data class EncryptedBlob(val nonceHex: String, val ciphertextHex: String)

    fun encryptRecord(jsonBytes: ByteArray, dek: ByteArray): EncryptedBlob {
        val nonce = randomBytes(NONCE_LEN)
        val ct = aesGcmEncrypt(jsonBytes, dek, nonce)
        return EncryptedBlob(toHex(nonce), toHex(ct))
    }

    fun decryptRecord(blob: EncryptedBlob, dek: ByteArray): ByteArray =
        aesGcmDecrypt(fromHex(blob.ciphertextHex), dek, fromHex(blob.nonceHex))
}
